tests/test_pandas.py

```diff
@@ -316,3 +316,100 @@ def test_eval():
         ),
         pd.DataFrame,
     )
+
+
+def test_crosstab_args() -> None:
+    a = [1, 2, 3, 4, 1, 2, 3, 4, 1, 2]
+    b: list = [4, 5, 6, 3, 4, 3, 5, 6, 5, 5]
+    c = [1, 3, 2, 3, 1, 2, 3, 1, 3, 2]
+    check(assert_type(pd.crosstab(a, b), pd.DataFrame), pd.DataFrame)
+    check(assert_type(pd.crosstab(a, [b, c]), pd.DataFrame), pd.DataFrame)
+    check(
+        assert_type(pd.crosstab(np.array(a), np.array(b)), pd.DataFrame), pd.DataFrame
+    )
+    check(
+        assert_type(pd.crosstab(np.array(a), [np.array(b), np.array(c)]), pd.DataFrame),
+        pd.DataFrame,
+    )
+    check(
+        assert_type(pd.crosstab(pd.Series(a), pd.Series(b)), pd.DataFrame), pd.DataFrame
+    )
+    check(
+        assert_type(
+            pd.crosstab(pd.Series(a), [pd.Series(b), pd.Series(c)]), pd.DataFrame
+        ),
+        pd.DataFrame,
+    )
+    values = [1, 2, 3, 4, 5, 6, 7, 8, 9, 10]
+    check(
+        assert_type(pd.crosstab(a, b, values=values, aggfunc=np.sum), pd.DataFrame),
+        pd.DataFrame,
+    )
+    check(
+        assert_type(
+            pd.crosstab(a, b, values=np.array(values), aggfunc=np.sum), pd.DataFrame
+        ),
+        pd.DataFrame,
+    )
+
+    check(
+        assert_type(
+            pd.crosstab(a, b, values=pd.Series(values), aggfunc=np.sum), pd.DataFrame
+        ),
+        pd.DataFrame,
+    )
+    check(
+        assert_type(pd.crosstab(a, b, values=values, aggfunc=np.mean), pd.DataFrame),
+        pd.DataFrame,
+    )
+    check(
+        assert_type(pd.crosstab(a, b, values=values, aggfunc="mean"), pd.DataFrame),
+        pd.DataFrame,
+    )
+
+    def m(x: pd.Series) -> float:
+        return x.sum() / len(x)
+
+    check(
+        assert_type(pd.crosstab(a, b, values=values, aggfunc=m), pd.DataFrame),
+        pd.DataFrame,
+    )
+
+    def m2(x: pd.Series) -> int:
+        return int(x.sum())
+
+    check(
+        assert_type(pd.crosstab(a, b, values=values, aggfunc=m2), pd.DataFrame),
+        pd.DataFrame,
+    )
+    check(
+        assert_type(
+            pd.crosstab(a, b, margins=True, margins_name="something"), pd.DataFrame
+        ),
+        pd.DataFrame,
+    )
+    check(
+        assert_type(pd.crosstab(a, b, margins=True, dropna=True), pd.DataFrame),
+        pd.DataFrame,
+    )
+    check(
+        assert_type(pd.crosstab(a, b, colnames=["a"], rownames=["b"]), pd.DataFrame),
+        pd.DataFrame,
+    )
+    rownames: list[tuple] = [("b", 1)]
+    colnames: list[tuple] = [("a",)]
+    # pyright can't figure out that these are hashable
```

Comment:
The issue here is that you have two arguments that are `list[HashableT]`, and `HashableT` is bound, so what happens is that one of them gets bound to a specific type, and the other one is different.
Consider this example:
```python
from typing import Hashable, TypeVar

HashableT = TypeVar("HashableT", bound=Hashable)
HashableT1 = TypeVar("HashableT1", bound=Hashable)
HashableT2 = TypeVar("HashableT2", bound=Hashable)


rownames: list[tuple] = [("b", 1)]
colnames: list[tuple] = [("a",)]

def f1(rn: list[HashableT], cn: list[HashableT]):
    pass

f1(rownames, colnames)


def f2(rn: list[HashableT1], cn: list[HashableT2]):
    pass

f2(rownames, colnames)
```
In this case, the call to `f1` fails because `rownames` and `colnames` are both `list[HashableT]`, but the call to `f2` works because the bindings are different, even though the `TypeVar` declarations are identical.

Reply:
Why would the type in the HashableT need to be the same?

Reply:
I mostly get it.  Clearly if you have f(x:T)->T: it has to be the same T, so something like f(a:list[T},b:list[T])->None also needs the two T to be the same.

Reply:
> Why would the type in the HashableT need to be the same?

Not entirely sure, but I think it has to do with the `bound=Hashable` declaration.  If you have a function `g(x: HashableT) -> HashableT`, and then do `result = g("abc")` followed by `reveal_type(result)`, you will get `str`, because pyright said "you passed a string, which fits the expected type, but now I know you are returning a string, and not any other Hashable". 

The same thing is happening here.  You bound one argument, so it expects the second argument to have the same binding.

At least that is what I *think* is going on.

Reply:
> I mostly get it. Clearly if you have f(x:T)->T: it has to be the same T, so something like f(a:list[T},b:list[T])->None also needs the two T to be the same.

Yes, that's how I see it as well.

Reply:
Fixed now.  

I think this is a wide spread problem with use of HashableT.  Pretty must any function that has 2 or more HashableT should have distinct HashableTs (e.g., HashableT0, HashableT1, ...)

Reply:
> Fixed now.
> 
> I think this is a wide spread problem with use of HashableT. Pretty must any function that has 2 or more HashableT should have distinct HashableTs (e.g., HashableT0, HashableT1, ...)

I think @twoertwein knows more about this.  I recall some discussion we had about this in the past

Reply:
As far as I understand, the primary purpose of TypeVars is binding types (whether it is between input and output or only between inputs). We often use TypeVars as a workaround to get invariant(? - always mix those up) behaviors for generic classes such as list, dict, Mapping and so on.